### PR TITLE
Exporter/Stats/Stackdriver: handle Summary Type

### LINKED ISSUE #1514: Decide how to handle Summary Type inside Stackdriver exporter
https://github.com/census-instrumentation/opencensus-java/pull/1501/files#diff-d2886f5b11a5e3be7a00a8f5fff5a1f0R114

### PR DESCRIPTION
**Pull request by mayurkale22**
Fixes #1514

### CODE REVIEW

**Comment by bogdandrutu** [comment]
One options is to have two timeseries for a SUMMARY (e.g. name = queue_size):
1. {name = "queue_size", label_key = "field",  type = "cumulative_double"}
    * label_value = "count" -> the count field in the SUMMARY.
    * label_value = "sum" -> the sum field in the SUMMARY.
2. {name = "queue_size_snapshot", label_key = "field",  type = "gauge_double"}
    * label_value = "count" -> the count field in the SNAPSHOT.
    * label_value = "sum" -> the sum field in the SNAPSHOT.
    * label_value = "pXX" -> the value at percentile XX.

Any thoughts?

**Comment by mayurkale22** [comment]
I have created a doc to discuss this with the Stackdriver team. 

(tl;dr)
**Option 2:**
1. {name = "queue_size",  type = "cumulative_distribution"}
    * distribution_count = "count" -> the count field in the SUMMARY
    * distribution_mean = "sum/count" -> the sum field in the SUMMARY.
    * Distribution_buckets = Without any buckets.
2. {name = "queue_size_snapshot", label_key = "percentile/quantile", type = "gauge_double"}
    * label_value = "pXX" -> the value at percentile XX.
    * In this case, we are loosing/ignoring snapshot->count and snapshot->mean.

**Comment by migibert** [comment]
Sounds good, I did not check on stackdriver side but why isn't the distribution type possible? It allows median, deciles, and some other stats

**Comment by bogdandrutu** [comment]
@migibert the SD distribution uses histograms and not quantiles so we cannot export already calculated percentiles to the SD distribution. Does this answer your question?

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java [hunk 1]
```diff
@@ -273,4 +306,103 @@ static String getDisplayNamePrefix(@javax.annotation.Nullable String metricNameP
       return metricNamePrefix;
     }
   }
+
+  @VisibleForTesting
+  static List<Metric> summaryMetricToDoubleGaugeMetric(Metric summaryMetric) {
```
Comment (bogdandrutu) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java]: Why is not in the Util as all the others?

Comment (mayurkale22) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java]: Done,

**Comment by migibert** [comment]
Perfectly, thanks for the explanation

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java [hunk 2]
```diff
@@ -172,8 +173,21 @@ void export() {
...
+          List<Metric> summaryToGaugeMetrics =
+              StackdriverExportUtils.summaryMetricToDoubleGaugeMetric(metric);
+          metricsList.ensureCapacity(metricsList.size() + summaryToGaugeMetrics.size());
+          for (Metric summaryToGaugeMetric : summaryToGaugeMetrics) {
+            if (registerMetricDescriptor(summaryToGaugeMetric.getMetricDescriptor())) {
+              metricsList.add(summaryToGaugeMetric);
+            }
+          }
+        } else {
+          if (registerMetricDescriptor(metric.getMetricDescriptor())) {
```
Comment (bogdandrutu) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java]: s/metric.getMetricDescriptor()/metricDescriptor?

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java [hunk 3]
```diff
@@ -172,8 +173,21 @@ void export() {
       Collection<Metric> producerMetricsList = metricProducer.getMetrics();
       metricsList.ensureCapacity(metricsList.size() + producerMetricsList.size());
       for (Metric metric : producerMetricsList) {
-        if (registerMetricDescriptor(metric.getMetricDescriptor())) {
-          metricsList.add(metric);
+        final io.opencensus.metrics.export.MetricDescriptor metricDescriptor =
+            metric.getMetricDescriptor();
+        if (metricDescriptor.getType() == Type.SUMMARY) {
+          List<Metric> summaryToGaugeMetrics =
+              StackdriverExportUtils.summaryMetricToDoubleGaugeMetric(metric);
```
Comment (bogdandrutu) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java]: Why do we have to do this every time? Can we cache the result?

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 4]
```diff
@@ -425,6 +434,114 @@ static void setResourceForBuilder(
     throw new IllegalArgumentException("Unknown subclass of MonitoredResource.");
   }
 
+  @VisibleForTesting
+  static List<io.opencensus.metrics.export.Metric> summaryMetricToDoubleGaugeMetric(
```
Comment (bogdandrutu) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: I think we need to split the creation of the metric descriptors from the creation of the timeseries. This logic is too complicated and we need to simplify it.

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java [hunk 5]
```diff
@@ -172,8 +173,21 @@ void export() {
...
+          metricsList.ensureCapacity(metricsList.size() + summaryToGaugeMetrics.size());
+          for (Metric summaryToGaugeMetric : summaryToGaugeMetrics) {
```
Comment (bogdandrutu) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterWorker.java]: Not necessary gauge, we have cumulative metrics as well.

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 6]
```diff
@@ -425,6 +434,143 @@ static void setResourceForBuilder(
...
+                },
+                Functions.<Void>returnNull());
+      }
+    }
+
+    // Metric for summary->count.
+    if (summaryCountTimeSeries.size() > 0) {
+      addMetric(
+          metricsList,
+          createNewMetricDescriptor(
```
Comment (bogdandrutu) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: This metric has alway the default unit "1". The sum and percentiles will have the metric unit.

**Comment by mayurkale22** [comment]
@bogdandrutu Thanks for the reviews.

**Pull request merged**